src/compiler/checker.ts

```diff
@@ -7684,8 +7682,11 @@ namespace ts {
                     return baseIndexedAccess && baseIndexedAccess !== errorType ? getBaseConstraint(baseIndexedAccess) : undefined;
                 }
                 if (t.flags & TypeFlags.Conditional) {
-                    const constraint = getConstraintOfConditionalType(<ConditionalType>t);
-                    return constraint && getBaseConstraint(constraint);
+                    const constraint = getConstraintFromConditionalType(<ConditionalType>t);
+                    constraintDepth++; // Penalize repeating conditional types (this captures the recursion within getConstraintFromConditionalType and carries it forward)

```

Comment:
Why the extra bump? Not that I mind, just curious.

Reply:
As I mentioned in the OP, the new circularity checking isn't quite enough to stop the crashes in `infiniteConstraints` - bottoming out the base constraint at depth 50 just takes too long, since `getConstraintFromConditionalType` is recursive into `getConstraintOfType` (in the distributive case), except it doesn't return the constraint it finds - it returns a conditional distributing over that constraint. In effect, we get an instance there, dive to depth 50, unwind, inject it into a new instance, then dive down to 50 again. The added depth just penalizes this pattern so it runs out of gas faster. Once the distributive constraint is gone (since it just simulates generic negations), we might not need it.